### ISSUE TITLE
Future of `string_processing`

### ISSUE THREAD

**Issue by JelleZijlstra** [psf/black#4208]
We merged string processing in May 2020 (#1132), almost four years ago. A few months later, I put the feature behind a flag because there were too many related bugs (#1609). Four years later, we're still in basically the same state: most of the open issues in https://github.com/psf/black/labels/C%3A%20preview%20style are about string processing.

I'm also increasingly convinced that string_processing's core feature, splitting up long strings, would be exceedingly controversial if we ever did turn it on by default. It is just too hard for a tool to figure out how to split a string in a way that makes sense to readers of the code. However, there are other things that string_processing does, and some of those are valuable and could perhaps be split off.

So my proposal is:
- Give up on splitting strings automatically; Black will simply not do that
- Split out any other useful features from string-processing into separate preview features that we can reasonably hope to enable in next year's stable style.

If you're interested in getting any part of string_processing into stable Black, please contribute a PR moving it to a separate preview feature.

I ran string_processing on my work codebase and noticed a few specific changes that could be split out. I'll list them in a separate comment.

**Comment by JelleZijlstra** [comment]
Here are some examples of changes that string-processing makes in my codebase that I'd consider accepting in a standalone preview feature:

* Add parentheses around a string so it fits in the line limit
```
         xxxxxx = xx.xxxxxxx.XXXXXxxxxXxxXxfxxxxxx(
-            xxxxx_xxxxxxx_x3_xxxx_xxxxxxxx=f"x3://{XXX_X3_XXXXXX}/{XXXX_XXXXXXX_X3_XXXX_XXXXXXXX}",
+            xxxxx_xxxxxxx_x3_xxxx_xxxxxxxx=(
+                f"x3://{XXX_X3_XXXXXX}/{XXXX_XXXXXXX_X3_XXXX_XXXXXXXX}"
+            ),
             xxxxx_xxx=[
```
* Merge implicitly concatenated strings if the result fits in the line limit
```
     xxx_xxxxxxx_xxxxxxxx = (
-        "XXX(XX(xx XXXXXXX '{xxxxx_xx}' XXX '{xxx_xx}', {xxx}, 0)) "
-        "XX {xxx}_x{xxxxxx}"
+        "XXX(XX(xx XXXXXXX '{xxxxx_xx}' XXX '{xxx_xx}', {xxx}, 0)) XX {xxx}_x{xxxxxx}"
     )
```
* Combination of both, looks much nicer
```
     name_to_base_expression = {
-        "view": f"CASE WHEN (playtime >= {VIEW_THRESHOLD} "
-        "or has_unmute) THEN 1 ELSE 0 END",
+        "view": (
+            f"CASE WHEN (playtime >= {VIEW_THRESHOLD} or has_unmute) THEN 1 ELSE 0 END"
+        ),
     }
```
* Remove `f` from parts of an implicitly concatenated f-string that don't have placeholders
```
             [
-                f"xxxxxx /xxxx/xxxxxxx/xxxx_xxxxxxx_xxxxx_xx.xx "
+                "xxxxxx /xxxx/xxxxxxx/xxxx_xxxxxxx_xxxxx_xx.xx "
                 f"--xxxxxxxx_xxxx {xxxxx_xxxxxxx_xxxxx} "
                 f"--xxxxxx_xxx {xxxxx_xxxxxx_xxxx} "
-                f"--xx_xxxx "
+                "--xx_xxxx "
                 f" {xxxxxx_xxxxxxx_xxxxx_xxx} "
                 f" {xx_xxx_fxxxx} "
             ]
```

**Comment by zanieb** [comment]
My 2-cents: I switched to Black's preview mode just for this feature. Otherwise, string processing is the vast majority of the time I spend manually formatting code. I would love to see this reach a place where it can be stabilized. I never had any problems with it, although I do understand there are _many_ edge cases. I think splitting it into individual features makes a lot of sense so you can make incremental progress. I find splitting long strings really valuable, but understand it's a tough problem — we've been hesitant to tackle it in Ruff as well.

**Comment by carschandler** [comment]
Agreed with @zanieb this feature is the main reason I use Black over Ruff. Thinking about splitting all of my strings manually gives me Java-scented nightmares.

**Comment by gdlxn-ibm** [comment]
I too consider string formatting an important feature / advantage of Black formatting.

**Comment by cobaltt7** [comment]
The `multiline_string_handling` docs currently say it combines strings like these:
```py
s = (
    "An "
    "implicit "
    "multiline "
    "string"
)

s = (  # Top comment
    "An "
    "implicit "
    "multiline "
    "string"
    # Bottom comment
)
``` 
However, that functionality is actually implemented by `string_processing`. But the `string_processing` docs say "User-made splits are respected when they do not exceed the line length limit." IMO concatenation is good, but strings should be force-split on newlines (see #1467). Then again, that might not be the best solution if we decide to move away from splitting long strings

**Comment by steve-mavens** [comment]
I suppose this is heresy, but is there a case here for a rule that's optional by design? This feature has been in "unstable" for years, because enough users don't want it, but there are also users here who do want this rule as it is, even without further improvement.

This seems to be a case where black can only do a "meh" job because the problem is hard.

I speculate that those with lots of long strings their programs[*] might well vastly prefer a "meh" job to no help at all, whereas those with only a few long strings would vastly prefer to split them manually (or just exceed the line-length) rather than tolerate "meh".

Unless/until black can do a "good" job, it seems to me that black's opinionated view is currently wrong (because it's only meh), but still right enough for a significant group of users. So, should black be configurable according to which kind of user you are, without having to call a feature "unstable" when actually it's as complete as any other?

[*] especially those for whom short lines are vital because they read code on a teletype terminal/mobile in portrait/cluttered IDE with little space for the source/3+ way diff.

**Comment by cobaltt7** [comment]
It's unstable because many of the issues in related to it are not acceptable to ship as a preview feature, even with the expectation that it only does a "meh" job. I.e, crashes, unstable formatting, modifying or duplicating comments, performance regressions
